fedbiomed/researcher/job.py

```diff
@@ -475,22 +494,131 @@ def start_nodes_training_round(self,
-                               'params': params,
-                               'optimizer_args': optimizer_args,
-                               'sample_size': m["sample_size"],
-                               'encryption_factor': encryption_factor,
-                               'timing': timing})
-
-                self._training_replies[round_].append(r)
+                response = Responses({
+                    'success': m['success'],
+                    'msg': m['msg'],
+                    'dataset_id': m['dataset_id'],
+                    'node_id': m['node_id'],
+                    'params_path': params_path,
+                    'params': params,
+                    'optimizer_args': optimizer_args,
+                    'optim_aux_var': optim_aux_var,
+                    'sample_size': m["sample_size"],
+                    'encryption_factor': encryption_factor,
+                    'timing': timing,
+                })
+                self._training_replies[round_].append(response)
 
         # return the list of nodes which answered because nodes in error have been removed
         return self._nodes
 
+    def upload_agg_optimizer_aux_var(
+        self,
+        aux_var: Dict[str, Dict[str, Any]],
+    ) -> Tuple[Optional[str], Dict[uuid.UUID, str]]:
+        """Upload auxiliary variables emitted by a researcher-side Optimizer.
+
+        Args:
+            aux_var: Dict of auxiliary variables emitted by an Optimizer held
+                by the researcher, that are to be uploaded after having been
+                structured into multiple files, to avoid information leakage
+                as well as content redundancy.
+
+        Returns:
+            url_shared: url of a file containing auxiliary variables shared
+                across all nodes, or None (in the absence of such information).
+            url_bynode: dict mapping urls of files containing node-specific
+                auxiliary variables to the nodes' id (a missing `nodes` key
+                indicates that this node has no such information to receive).
+        """
+        # Split the information between shared and node-wise dictionaries.
+        aux_shared, aux_bynode = self._prepare_agg_optimizer_aux_var(
+            aux_var=aux_var, nodes=list(self._nodes)
+        )
+        # Upload the shared information that all nodes will download.
+        if aux_shared:
+            path = os.path.join(
+                self._keep_files_dir, f"aux_var_shared_{uuid.uuid4()}.mpk"
+            )
+            Serializer.dump(aux_shared, path)
+            url_shared = self.repo.upload_file(path)["file"]
+        else:
+            url_shared = None
+        # Upload the node-specific information, with node-specific urls.
+        url_bynode = {}  # type: Dict[uuid.UUID, str]
+        for node_id, node_aux in aux_bynode.items():
+            if not node_aux:
+                continue
+            path = os.path.join(
```

Comment:
I dont really understand that part: what if `node_aux = {}`? should it be considered as an error instead? Or do you see any specific setting where it can happen